In [6]:
def countSetBits(n):
    count = 0
    while n:
        n  = n & (n-1)
        print(n)
        count += 1
    return count

n = 9
print(countSetBits(n))
print(bin(9))
# print(bin(8) , bin(7))
# print(8 & 7)


8
0
2
0b1001


In [1]:
def bit_operations_demo(n, i):
    print(f"\n=== Bitwise Operations on n = {n} (binary: {bin(n)[2:]}) ===\n")

    # 1. Get ith bit
    get_bit = (n >> i) & 1
    print(f"1. Get {i}th bit           : (n >> i) & 1       = {get_bit}")

    # 2. Set ith bit to 1
    set_bit = n | (1 << i)
    print(f"2. Set {i}th bit to 1      : n | (1 << i)       = {set_bit} (binary: {bin(set_bit)[2:]})")

    # 3. Clear ith bit (set to 0)
    clear_bit = n & ~(1 << i)
    print(f"3. Clear {i}th bit to 0    : n & ~(1 << i)      = {clear_bit} (binary: {bin(clear_bit)[2:]})")

    # 4. Toggle ith bit
    toggle_bit = n ^ (1 << i)
    print(f"4. Toggle {i}th bit        : n ^ (1 << i)       = {toggle_bit} (binary: {bin(toggle_bit)[2:]})")

    # 5. Check if n is power of 2 (only 1 bit set)
    is_power_of_2 = n > 0 and (n & (n - 1)) == 0
    print(f"5. Is power of 2           : n & (n-1) == 0     = {is_power_of_2}")

    # 6. Count number of set bits (Brian Kernighan's algo)
    count = 0
    temp = n
    while temp:
        temp = temp & (temp - 1)
        count += 1
    print(f"6. Count set bits          : while n: n &= n-1  = {count}")

    # 7. Remove the last set bit
    remove_last = n & (n - 1)
    print(f"7. Remove last set bit     : n & (n - 1)        = {remove_last} (binary: {bin(remove_last)[2:]})")

    # 8. Get only the last set bit
    last_set = n & -n
    print(f"8. Get only last set bit   : n & -n             = {last_set} (binary: {bin(last_set)[2:]})")

    # 9. Turn ON lowest 0 bit
    turn_on_lowest_0 = n | (n + 1)
    print(f"9. Turn on lowest 0 bit    : n | (n + 1)        = {turn_on_lowest_0} (binary: {bin(turn_on_lowest_0)[2:]})")

    # 10. Complement bits in range
    bit_len = n.bit_length()
    mask = (1 << bit_len) - 1
    complement = n ^ mask
    print(f"10. Bitwise complement     : n ^ ((1<<bits)-1)  = {complement} (binary: {bin(complement)[2:]})")


# Run the demo for a sample number
bit_operations_demo(70, 2)



=== Bitwise Operations on n = 70 (binary: 1000110) ===

1. Get 2th bit           : (n >> i) & 1       = 1
2. Set 2th bit to 1      : n | (1 << i)       = 70 (binary: 1000110)
3. Clear 2th bit to 0    : n & ~(1 << i)      = 66 (binary: 1000010)
4. Toggle 2th bit        : n ^ (1 << i)       = 66 (binary: 1000010)
5. Is power of 2           : n & (n-1) == 0     = False
6. Count set bits          : while n: n &= n-1  = 3
7. Remove last set bit     : n & (n - 1)        = 68 (binary: 1000100)
8. Get only last set bit   : n & -n             = 2 (binary: 10)
9. Turn on lowest 0 bit    : n | (n + 1)        = 71 (binary: 1000111)
10. Bitwise complement     : n ^ ((1<<bits)-1)  = 57 (binary: 111001)


📌 1. Why Bit Manipulation?
Bit manipulation uses bitwise operators to perform operations at the bit level, enabling:

🔸 Constant-time operations (very fast)

🔸 Space-efficient coding

🔸 Solving problems related to sets, power-of-2, parity, etc


| Operator    | Symbol | Meaning                           | Example (a=5, b=3) |     |         |
| ----------- | ------ | --------------------------------- | ------------------ | --- | ------- |
| AND         | `&`    | 1 if both bits are 1              | `5 & 3 = 1`        |     |         |
| OR          | \`     | \`                                | 1 if any bit is 1  | \`5 | 3 = 7\` |
| XOR         | `^`    | 1 if bits are different           | `5 ^ 3 = 6`        |     |         |
| NOT         | `~`    | Flip all bits (1's comp.)         | `~5 = -6`          |     |         |
| LEFT SHIFT  | `<<`   | Shift bits left (multiply by 2^n) | `5 << 1 = 10`      |     |         |
| RIGHT SHIFT | `>>`   | Shift bits right (divide by 2^n)  | `5 >> 1 = 2`       |     |         |


In [8]:
n = 6
# 7 % 2 = 1
# 3 % 2 = 1
# 1 % 2 = 1
res = ""
while n:
    res += str(n % 2)
    n = n // 2

print(res[::-1])
    

110


In [ ]:
# Input: 70 3
# Output: 1 70 66
# Explanation: Bit at the 3rd position from LSB is 1. (1 0 0 0 1 1 0) .The value of the given number after setting the 3rd bit is 70. The value of the given number after clearing 3rd bit is 66. (1 0 0 0 0 1 0)
# Input: 8 1
# Output: 0 9 8
# Explanation: Bit at the first position from LSB is 0. (1 0 0 0)  .The value of the given number after setting the 1st bit is 9. (1 0 0 1).  The value of the given number after clearing 1st bit is 8. (1 0 0 0)

# 70 = 1  0  0  0 1 1 0
#      64 32 16 8 4 2 0
# 8 = 1000
# 9 = 1001
# print(bin(70)[2:])
# print(int('1010010',2))

n = 70
pos = 3
bit = (n >> (pos - 1)) & 1
set_bit = n | (1 << (pos - 1))
clear_bit = n & ~(1 << (pos - 1))





1


| Task             | Code                     |            |
| ---------------- | ------------------------ | ---------- |
| Get ith bit      | `(n >> i) & 1`           |            |
| Set ith bit      | \`n                      | (1 << i)\` |
| Clear ith bit    | `n & ~(1 << i)`          |            |
| Toggle ith bit   | `n ^ (1 << i)`           |            |
| Check power of 2 | `n > 0 and n & (n-1)==0` |            |
| Count 1s         | `while n: n &= n-1`      |            |


In [6]:

# 476. Number Complement
# Solved
# Easy
# Topics
# premium lock icon
# Companies
# The complement of an integer is the integer you get when you flip all the 0's to 1's and all the 1's to 0's in its binary representation.

# For example, The integer 5 is "101" in binary and its complement is "010" which is the integer 2.
# Given an integer num, return its complement.

 

# Example 1:

# Input: num = 5
# Output: 2
# Explanation: The binary representation of 5 is 101 (no leading zero bits), and its complement is 010. So you need to output 2.
# Example 2:

# Input: num = 1
# Output: 0
# Explanation: The binary representation of 1 is 1 (no leading zero bits), and its complement is 0. So you need to output 0.


class Solution:
    def findComplement(self, num: int) -> int:
        bit_len = num.bit_length()
        mask = (1 << bit_len) - 1
        return num ^ mask

sol = Solution()
print(sol.findComplement(5))

2


| Task                        | Shortcut                          |           |
| --------------------------- | --------------------------------- | --------- |
| Power of 2                  | `n > 0 and n & (n - 1) == 0`      |           |
| Remove last set bit         | `n & (n - 1)`                     |           |
| Get last set bit            | `n & (-n)`                        |           |
| Turn on lowest unset bit    | \`n                               | (n + 1)\` |
| Check ith bit is set        | `n & (1 << i)`                    |           |
| Flip all bits in used range | `n ^ ((1 << n.bit_length()) - 1)` |           |


| Operation        | Code                | Meaning                         |                    |
| ---------------- | ------------------- | ------------------------------- | ------------------ |
| Get ith bit      | `(n >> i) & 1`      | Check if ith bit is 1           |                    |
| Set ith bit      | \`n                 | (1 << i)\`                      | Force ith bit to 1 |
| Clear ith bit    | `n & ~(1 << i)`     | Force ith bit to 0              |                    |
| Toggle ith bit   | `n ^ (1 << i)`      | Flip ith bit                    |                    |
| Count 1s         | `while n: n &= n-1` | Fastest way to count 1s         |                    |
| Check power of 2 | `n & (n - 1) == 0`  | Only powers of 2 have 1 set bit |                    |


| Trick                             | Code                                            | Use Case                        |
| --------------------------------- | ----------------------------------------------- | ------------------------------- |
| **Check if number is power of 2** | `n & (n - 1) == 0`                              | Only 1 bit is set in power of 2 |
| **Count set bits**                | `bin(n).count('1')` or `Brian Kernighan’s Algo` | Count how many bits are 1       |
| **Remove rightmost set bit**      | `n & (n - 1)`                                   | Useful in loops                 |
| **Get rightmost set bit**         | `n & (-n)`                                      | Isolate the last `1`            |
| **Swap two numbers without temp** | `a ^= b; b ^= a; a ^= b`                        | Space-saving trick              |
| **Clear LSB upto i-th bit**       | `n & (~((1 << (i+1)) - 1))`                     | Clear bits from 0 to i          |
| **Clear MSB from i-th bit**       | `n & ((1 << i) - 1)`                            | Keep only lower i bits          |


In [5]:
print((70 >> 1) & 1)

print(70 | (1 << 2))

print(70 & ~(1 << 2))

1
70
66


In [7]:
#kth bit is set or not 
n = 4
k = 2
print ((n >> k) & 1)

1


1000110
0000001  ← 1 is mask for last bit
-------
0000000 → result is 0 → even

1000111
0000001
-------
0000001 → result is 1 → odd


In [16]:
#odd or even without using %2
class Solution:
    def isEven (self, n):
        return False if (n & 1) else True
        # code here 
print(Solution.isEven(Solution, 3))        

False


In [30]:
# set right most bit to 1 
#n = 6 The binary representation of 6 is 110. After setting right most bit it becomes 111 which is 7.
n = 6
print (n | n + 1)

7


In [32]:
# 476. Number Complement
# Solved
# Easy
# Topics
# premium lock icon
# Companies
# The complement of an integer is the integer you get when you flip all the 0's to 1's and all the 1's to 0's in its binary representation.

# For example, The integer 5 is "101" in binary and its complement is "010" which is the integer 2.
# Given an integer num, return its complement.

 

# Example 1:

# Input: num = 5
# Output: 2
# Explanation: The binary representation of 5 is 101 (no leading zero bits), and its complement is 010. So you need to output 2.
# Example 2:

# Input: num = 1
# Output: 0
# Explanation: The binary representation of 1 is 1 (no leading zero bits), and its complement is 0. So you need to output 0.
 

num = 5
print (num ^ (1 << num.bit_length()) - 1)

2


In [36]:
# 191. Number of 1 Bits
# Easy
# Topics
# premium lock icon
# Companies
# Given a positive integer n, write a function that returns the number of set bits in its binary representation (also known as the Hamming weight).

 

# Example 1:

# Input: n = 11

# Output: 3

# Explanation:

# The input binary string 1011 has a total of three set bits.

# Example 2:

# Input: n = 128

# Output: 1

# Explanation:

# The input binary string 10000000 has a total of one set bit.

# Example 3:

# Input: n = 2147483645

# Output: 30

# Explanation:

# The input binary string 1111111111111111111111111111101 has a total of thirty set bits.

 

n = 11
# 1011 ans = 0100
print(bin(n)[2:].count('1'))

# using Brian Kernighan's Algorithm
count  = 0
while n : 
    n &= n - 1
    count += 1
count 


3


3

In [50]:
# 338. Counting Bits
# Easy
# Topics
# premium lock icon
# Companies
# Hint
# Given an integer n, return an array ans of length n + 1 such that for each i (0 <= i <= n), ans[i] is the number of 1's in the binary representation of i.

 

# Example 1:

# Input: n = 2
# Output: [0,1,1]
# Explanation:
# 0 --> 0
# 1 --> 1
# 2 --> 10
# Example 2:

# Input: n = 5
# Output: [0,1,1,2,1,2]
# Explanation:
# 0 --> 0
# 1 --> 1
# 2 --> 10
# 3 --> 11
# 4 --> 100
# 5 --> 101
 

# Constraints:

# 0 <= n <= 105
 

n = 5
res = []
for i in range(n + 1):
    res.append(bin(i)[2:].count('1'))
res    

[0, 1, 1, 2, 1, 2]

In [3]:
# Input: n = 4
# Output: 5
# Explanation: For numbers from 1 to 4. For 1: 0 0 1 = 1 set bits For 2: 0 1 0 = 1 set bits For 3: 0 1 1 = 2 set bits For 4: 1 0 0 = 1 set bits Therefore, the total set bits is 5.

# n = 4
# res = []
# for i in range(n + 1):
#     res.append(bin(i)[2:].count('1'))
# sum(res)    

n = 4
n


4

In [1]:
# 190. Reverse Bits
# Easy
# Topics
# premium lock icon
# Companies
# Reverse bits of a given 32 bits unsigned integer.

# Note:

# Note that in some languages, such as Java, there is no unsigned integer type. In this case, both input and output will be given as a signed integer type. They should not affect your implementation, as the integer's internal binary representation is the same, whether it is signed or unsigned.
# In Java, the compiler represents the signed integers using 2's complement notation. Therefore, in Example 2 above, the input represents the signed integer -3 and the output represents the signed integer -1073741825.
 

# Example 1:

# Input: n = 00000010100101000001111010011100
# Output:    964176192 (00111001011110000010100101000000)
# Explanation: The input binary string 00000010100101000001111010011100 represents the unsigned integer 43261596, so return 964176192 which its binary representation is 00111001011110000010100101000000.
# Example 2:

# Input: n = 11111111111111111111111111111101
# Output:   3221225471 (10111111111111111111111111111111)
# Explanation: The input binary string 11111111111111111111111111111101 represents the unsigned integer 4294967293, so return 3221225471 which its binary representation is 10111111111111111111111111111111.



class Solution:
    def reverseBits(self, n: int) -> int:
        res = 0
        for _ in range(32):
            last_bit = n & 1            # get last bit
            res = (res << 1) | last_bit # shift result and add last_bit
            n >>= 1                     # shift n right
        return res

sol = Solution()
print(sol.reverseBits(43261596))

964176192


In [10]:
from collections import Counter
s = "213123"
substrings = []
maxx = 1
for i in range(len(s)):
    for j in range(i + 1, len(s) + 1):
        substrings.append(s[i:j])
        hashmap = Counter(s[i:j])
        for item , val in hashmap.items():
             values = list(hashmap.values())
             ones = values.count(1)
             others = [v for v in values if v != 1]
             if ones == 1 and len(set(others)) == 1:
                maxx = max(len(s[i:j]) , maxx)
print(maxx)


5
